Implement class aware retrofit meter function

arhont375 opened the pull request:
Add new meter function class for retrofit clients that provide more information:
- Raw path from the Retrofit annotaiton
- Consistent service name based on class name, by default with `service` tag
- httpMethod and httpStatus

This changes does not replace previous RetrofitMeterIdFunction as it would be breaking changes and maybe some users will prefer other variant :thinking: 

In addition to this Tag names for metric may be adjusted as well to comply with Armeria server format (currently use `path` and `httpMethod` whether Armeria server metrics provide `route` that contains path and http method at the same time)
Closes #2345

claassistantio commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/line/armeria?pullRequest=2356) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/line/armeria?pullRequest=2356) before we can accept your contribution.<br/><hr/>**arhont375** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/line/armeria?pullRequest=2356) it.</sub>

skkap commented on retrofit2/src/main/java/com/linecorp/armeria/client/retrofit2/RetrofitClassAwareMeterIdPrefixFunction.java:
```diff
@@ -0,0 +1,179 @@
+            final Annotation[] declaredAnnotations = clientServiceMethod.getDeclaredAnnotations();
+            if (declaredAnnotations.length == 0) {
+                continue;
+            }
```
I wonder if this can be removed without affecting the logic and performance?

trustin replied:
+1 for removing for simplicity

skkap commented on retrofit2/src/main/java/com/linecorp/armeria/client/retrofit2/RetrofitClassAwareMeterIdPrefixFunction.java:
```diff
@@ -0,0 +1,179 @@
+ * </ul>
+ */
+public class RetrofitClassAwareMeterIdPrefixFunction implements MeterIdPrefixFunction {
+    private static final List<Class> retrofitAnnotations = ImmutableList.of(
```
It looks like in Armeria it is common to add an empty line here as you did in `RetrofitClassAwareMeterIdPrefixFunctionTest`.

trustin commented on retrofit2/src/main/java/com/linecorp/armeria/client/retrofit2/RetrofitClassAwareMeterIdPrefixFunction.java:
```diff
@@ -0,0 +1,179 @@
+            }
+        }
+
+        tagsBuilder.add(Tag.of("method", log.method().name()));
```
A meter must have same set of tags regardless of whether `Invocation` is available or not, because otherwise Prometheus will complain.

trustin commented on retrofit2/src/main/java/com/linecorp/armeria/client/retrofit2/RetrofitClassAwareMeterIdPrefixFunction.java:
```diff
@@ -0,0 +1,179 @@
+            } catch (NoSuchMethodException | IllegalAccessException | InvocationTargetException e) {
+                // Should never happen on valid Retrofit client.
+                throw new IllegalStateException("Provided Retrofit client have unexpected annotation: " +
+                                                annotation.annotationType(), e);
```
Isn't it better skipping rather than raising an exception? A user may have added non-Retrofit annotations for other purposes.

trustin replied:
Ah I see. Thanks for explanation.

trustin commented:
How about this:

- Make `RetrofitMeterIdPrefixFunctionBuilder.withServiceClass()` package-private to hide it from the public API
- Add `of(String name, Class serviceClass)` and `builder(String name, Class<?> serviceClass)` to `RetrofitMeterIdPrefixFunction`
  - and remove the factory methods in `RetrofitClassAwareMeterIdPrefixFunction`
- Make `RetrofitClassAwareMeterIdPrefixFunction` package-private to hide it from the public API

..?

trustin commented on retrofit2/src/main/java/com/linecorp/armeria/client/retrofit2/RetrofitMeterIdPrefixFunctionBuilder.java:
```diff
@@ -51,10 +54,23 @@ public RetrofitMeterIdPrefixFunctionBuilder withServiceTag(String serviceTagName
      */
-    public RetrofitMeterIdPrefixFunction build() {
-        return new RetrofitMeterIdPrefixFunction(name, serviceTagName, defaultServiceName);
+    public MeterIdPrefixFunction build() {
```
We can continue returning `RetrofitMeterIdPrefixFunction` if we make `RetrofitClassAwareMeterIdPrefixFunction` extend `RetrofitMeterIdPrefixFunction`. Let's make `RetrofitMeterIdPrefixFunction` non-final with package-private constructor and let `RetrofitClassAwareMeterIdPrefixFunction` extend it.

trustin commented on retrofit2/src/main/java/com/linecorp/armeria/client/retrofit2/RetrofitMeterIdPrefixFunction.java:
```diff
@@ -48,7 +48,7 @@
      * Returns a newly created {@link RetrofitMeterIdPrefixFunction} with the specified {@code name}.
      */
-    public static RetrofitMeterIdPrefixFunction of(String name) {
+    public static MeterIdPrefixFunction of(String name) {
```
We can continue returning `RetrofitMeterIdPrefixFunction` if we make `RetrofitClassAwareMeterIdPrefixFunction` extend `RetrofitMeterIdPrefixFunction`. Let's make `RetrofitMeterIdPrefixFunction` non-final with package-private constructor and let `RetrofitClassAwareMeterIdPrefixFunction` extend it.

trustin replied:
This is not fixed yet.

trustin commented:
@kojilin PTAL as well.

kojilin commented on retrofit2/src/main/java/com/linecorp/armeria/client/retrofit2/RetrofitClassAwareMeterIdPrefixFunction.java:
```diff
@@ -0,0 +1,179 @@
+ * <ul>
+ *     <li>{@code service(or serviceTagName} - Retrofit service interface name</li>
+ *     <li>{@code path}   - Retrofit service interface method path taken from method annotation</li>
+ *     <li>{@code method} - Retrofit service interface method</li>
```
Looks like this may fall back to http method name if `Invocation` is null?

arhont375 commented on retrofit2/src/main/java/com/linecorp/armeria/client/retrofit2/RetrofitClassAwareMeterIdPrefixFunction.java:
```diff
@@ -0,0 +1,190 @@
+
+        return ImmutableList.of(
+                Tag.of(HTTP_METHOD_TAG_NAME, httpMethod),
+                Tag.of(PATH_TAG_NAME, httpPath)
```
How about 
> This decorator currently use `path` and `httpMethod` labels whether Armeria server metrics provide `route` label that contains path and http method at the same time

Makes sense to unify? Or just add another label :thinking:  I feel that it maybe easier to make queries when path and httpMethod separated

trustin replied:
Agreed. Separate tags sounds good to me.

trustin commented on retrofit2/src/main/java/com/linecorp/armeria/client/retrofit2/RetrofitClassAwareMeterIdPrefixFunction.java:
```diff
@@ -0,0 +1,190 @@
+ * the {@link Tag}s derived from the {@link RequestLog} properties, {@link Invocation} and provided retrofit
+ * client.
+ * <ul>
+ *     <li>{@code service(or serviceTagName} - Retrofit service interface name</li>
```
`{@code service} (or {@code serviceTagName})`

trustin commented on retrofit2/src/main/java/com/linecorp/armeria/client/retrofit2/RetrofitClassAwareMeterIdPrefixFunction.java:
```diff
@@ -0,0 +1,190 @@
+ *                          or {@code UNKNOWN} if Retrofit service interface method available</li>
+ *     <li>{@code httpMethod} - HTTP method name from Retrofit service interface method annotation
+ *                          or from {@link RequestLog#method()} if Retrofit service interface
+ *                          method not available</li>
```
Incorrect indentation at the 2nd and 3rd line

trustin commented on retrofit2/src/main/java/com/linecorp/armeria/client/retrofit2/RetrofitMeterIdPrefixFunctionBuilder.java:
```diff
@@ -51,10 +53,24 @@ public RetrofitMeterIdPrefixFunctionBuilder withServiceTag(String serviceTagName
+     *
+     * @param serviceClass the class defining client with Retrofit.
+     */
+    public RetrofitMeterIdPrefixFunctionBuilder withServiceClass(Class<?> serviceClass) {
```
What do you think about making this method package-private and adding `RetrofitMeterIdPrefixFunction.builder(String, Class)` and `of(String, Class)` which sets the service class at construction time?

arhont375 replied:
If I understood your idea correctly we don't need this method at all if will provide new package private constructor.

arhont375 replied:
Ah, I understood the way I changed will not work because class package private.
At the same time it feels strange to me to add such builder methods to the `RetrofitMeteridPrefixFunction` class because as a result we would return another specialization.

I feel that it would be better to make `RetrofitClassAwareMeterIdPrefixFunction` class either public or provide method to build this class in the builder (actually revert my last changes and add one more method to builder).

What do you think? Will you prefer way that you suggested initially or add more methods to builder?

ikhoon commented on retrofit2/src/main/java/com/linecorp/armeria/client/retrofit2/RetrofitClassAwareMeterIdPrefixFunction.java:
```diff
@@ -0,0 +1,190 @@
+        return new MeterIdPrefix(name, tagListBuilder.build());
+    }
+
+    private void buildTags(ImmutableList.Builder<Tag> tagsBuilder, RequestLog log) {
```
How about changing `RetrofitMeterIdPrefixFunction.buildTags` to `protected`? 
And remove the overrided methods:
- `activeRequestPrefix(MeterRegistry, RequestLog)`
- `apply(MeterRegistry, RequestLog)`?

arhont375 replied:
Usage of overriding methods gives us ability for a bit of microoptimization by providing expected list size. That by claims from guava may work faster then unbounded immutable list builder.

trustin replied:
I guess it's fine as it is unless @ikhoon has a strong opinion on this.

ikhoon replied:
> Usage of overriding methods gives us ability for a bit of microoptimization by providing expected list size. That by claims from guava may work faster then unbounded immutable list builder.

Sorry I missed this message. 

> I guess it's fine as it is unless @ikhoon has a strong opinion on this.

I don't have strong opinion here. Let's leave it as it is. :-)

minwoox replied:
Oops, I found out that we don't use exact number for the expected size [here](https://github.com/line/armeria/blob/master/retrofit2/src/main/java/com/linecorp/armeria/client/retrofit2/RetrofitMeterIdPrefixFunction.java#L89).
Could you change that as well please?
On my second thought, perhaps we can set the `builderWithExpectedSize` with the number + 1 in `RetrofitMeterIdPrefixFunction`, so that we can remove `activeRequestPrefix()` and `apply()` in `RetrofitClassAwareMeterIdPrefixFunction` and just overriding `buildTags()`.
Either way is fine for me though. 😉

ikhoon commented on retrofit2/src/test/java/com/linecorp/armeria/client/retrofit2/RetrofitClassAwareMeterIdPrefixFunctionTest.java:
```diff
@@ -0,0 +1,241 @@
+import retrofit2.http.POST;
+import retrofit2.http.PUT;
+
+public class RetrofitClassAwareMeterIdPrefixFunctionTest {
```
nit
```suggestion 
class RetrofitClassAwareMeterIdPrefixFunctionTest {
```

ikhoon commented on retrofit2/src/test/java/com/linecorp/armeria/client/retrofit2/RetrofitClassAwareMeterIdPrefixFunctionTest.java:
```diff
@@ -0,0 +1,241 @@
+                                                                    .meterRegistry(meterRegistry)
+                                                                    .build();
+
+    interface Example {
```
nit
```suggestion
    private interface Example {
```

ikhoon commented on retrofit2/src/test/java/com/linecorp/armeria/client/retrofit2/RetrofitClassAwareMeterIdPrefixFunctionTest.java:
```diff
@@ -0,0 +1,241 @@
+                protected HttpResponse doTrace(ServiceRequestContext ctx, HttpRequest req) {
+                    return HttpResponse.of(HttpStatus.OK);
+                }
+            });
```
nit: 
1) If you want to handle all methods
```java
sb.service("/foo", (ctx, req) -> HttpResponse.of(HttpStatus.OK));
```
2) If you want to handle `GET`, `POST` and `TRACE` together
```java
sb.service(Route.builder().path("/foo")
                .methods(HttpMethod.GET, HttpMethod.POST, HttpMethod.TRACE)
                .build(),
           (ctx, req) -> HttpResponse.of(HttpStatus.OK));
```

minwoox commented on retrofit2/src/main/java/com/linecorp/armeria/client/retrofit2/RetrofitClassAwareMeterIdPrefixFunction.java:
```diff
@@ -0,0 +1,190 @@
+     * Returns a newly created {@link RetrofitClassAwareMeterIdPrefixFunction} with the specified {@code name}
+     * and {@code serviceClass}.
+     */
+    public static RetrofitMeterIdPrefixFunction of(String name, Class<?> serviceClass) {
```
A user cannot create this because `RetrofitClassAwareMeterIdPrefixFunction` is package-private. We should move this method upto`RetrofitMeterIdPrefixFunction`.

minwoox commented on retrofit2/src/main/java/com/linecorp/armeria/client/retrofit2/RetrofitClassAwareMeterIdPrefixFunction.java:
```diff
@@ -0,0 +1,190 @@
+     * Returns a newly created {@link RetrofitMeterIdPrefixFunctionBuilder} with the specified {@code name}
+     * and {@code serviceClass}.
+     */
+    public static RetrofitMeterIdPrefixFunctionBuilder builder(String name, Class<?> serviceClass) {
```
ditto

minwoox commented on retrofit2/src/main/java/com/linecorp/armeria/client/retrofit2/RetrofitClassAwareMeterIdPrefixFunction.java:
```diff
@@ -0,0 +1,190 @@
+
+        this.name = name;
+        this.serviceTagName = firstNonNull(serviceTagName, "service");
+        this.serviceName = serviceClass.getSimpleName();
```
`this.serviceName` -> `serviceName`

minwoox commented on retrofit2/src/main/java/com/linecorp/armeria/client/retrofit2/RetrofitMeterIdPrefixFunctionBuilder.java:
```diff
@@ -33,11 +33,18 @@
 
+    RetrofitMeterIdPrefixFunctionBuilder(String name, Class<?> serviceClass) {
+        this.name = name;
+        this.serviceClass = serviceClass;
```
We should make sure that both `serviceClass` and `defaultServiceName` are not set together.

arhont375 replied:
Maybe I'm wrong, but I don't think it would affect usability:
If user use this builder then it's already clear that he want to use `serviceClass` but he still able to modify `serviceTag` with `withServiceTag` method.

Maybe comment on `withServiceTag` will help on this to clarify?

trustin replied:
Isn't `defaultServiceName` unused at all when `serviceClass` is used? How about:

- throwing an `IllegalStateException("defaultServiceName cannot be set for a builder created with serviceClass.")` for `withServiceTag(String, String)`,
- and adding `withServiceTag(String serviceTagName)` which works for both builder types.
  - If created without serviceClass, simply delegate to `withServiceTag(serviceTagName, "UNKNOWN")`

..?

arhont375 replied:
Why not to split this method into 2 and mark original `withServiceTag(String, String)` as deprecated?
- `withServiceTag(String serviceTag)`
- `withServiceName(String serviceName)` <-- I was curious why we need to provide `defaultServiceName` that used only in cases when `Invocation` not found (see [code](https://github.com/line/armeria/blob/master/retrofit2/src/main/java/com/linecorp/armeria/client/retrofit2/RetrofitMeterIdPrefixFunction.java#L106)). Let's just provide ability to use client provided service name that would be used for all cases and don't need to take name from the `Invocation` log then

arhont375 replied:
So this code from the link above:
```java
        if (invocation != null) {
            if (serviceTagName != null) {
                tagListBuilder.add(Tag.of(serviceTagName,
                                          invocation.method().getDeclaringClass().getSimpleName()));
            }
            tagListBuilder.add(Tag.of("method", invocation.method().getName()));
        } else {
            if (serviceTagName != null) {
                assert defaultServiceName != null;
                tagListBuilder.add(Tag.of(serviceTagName, defaultServiceName));
            }
            tagListBuilder.add(Tag.of("method", log.method().name()));
        }
```
Would be simplified to:
```
        if (serviceTagName != null) {
               tagListBuilder.add(Tag.of(serviceTagName, serviceName));
        }
        if (invocation != null) {
            tagListBuilder.add(Tag.of("method", invocation.method().getName()));
        } else {
            tagListBuilder.add(Tag.of("method", log.method().name()));
        }
```

arhont375 replied:
Unfortunately we can't make this ^ change now as it will break current metrics (as it may happen that provided serviceName would be different), but I think it should be way to go in the future

trustin replied:
> Why not to split this method into 2 and mark original `withServiceTag(String, String)` as deprecated?

Sounds good to me!

minwoox replied:
How about deprecating the current `withServiceTag` and adding Two methods:
- `withServiceName(String serviceTagName, String defaultServiceName)`
  - We should mention when the `defaultServiceName` is used in Javadoc
    - `@param defaultServiceName the default value of the tag, e.g. {@code "myService"}. It's only used when {@link InvocationUtil#getInvocation(RequestLog)} returns {@code null}.`
- `withServiceCalss(String serviceTagName, Class<?> serviceClass)`

arhont375 replied:
I think it's better to separate into `withServiceName(serviceName)` and `withServiceTag(serviceTagName)` methods as we may want to override service name for the case when class used. Also we may just introduce meaningful defaults to reduce code for users.

minwoox replied:
> we may want to override service name for the case when class used

Ah, yes I missed that point. 😄 

So, now we can set the property one by one, how about removing the prefix `with`?
Also, What do you think adding a setter for the serviceClass so that a user can do:
```
RetrofitMeterIdPrefixFunction.of(name)
                             .serviceTag(serviceTagName)
                             .serviceClass(serviceClass) // I'm not sure if this is a good name though. ;)
                             .build()
```

trustin commented on retrofit2/src/main/java/com/linecorp/armeria/client/retrofit2/RetrofitClassAwareMeterIdPrefixFunction.java:
```diff
@@ -0,0 +1,190 @@
+    public static RetrofitMeterIdPrefixFunctionBuilder builder(String name, Class<?> serviceClass) {
+        return new RetrofitMeterIdPrefixFunctionBuilder(requireNonNull(name, "name"),
+                                                        requireNonNull(serviceClass, "serviceClass"));
+    }
```
These two classes could be removed?

trustin commented on retrofit2/src/main/java/com/linecorp/armeria/client/retrofit2/RetrofitClassAwareMeterIdPrefixFunction.java:
```diff
@@ -0,0 +1,172 @@
+
+        this.name = name;
+        this.serviceTagName = firstNonNull(serviceTagName, "service");
+        serviceName = serviceClass.getSimpleName();
```
Is it possible for a user to use an anonymous class as a service class? In that case, `getSimpleName` will return an empty string. Perhaps we should strip only the package names in that case?

arhont375 replied:
Is it possible to create Retrofit client with anonymous class?
May I ask to provide example?

trustin replied:
It seems like `Retrofit.create(Class)` requires the class to be an interface. I guess it's not possible to create a Retrofit client with an anonymous class, because an anonymous class is never an interface. :sweat_smile:

minwoox commented on retrofit2/src/main/java/com/linecorp/armeria/client/retrofit2/RetrofitMeterIdPrefixFunction.java:
```diff
@@ -52,34 +53,58 @@ public static RetrofitMeterIdPrefixFunction of(String name) {
     @Override
     public MeterIdPrefix activeRequestPrefix(MeterRegistry registry, RequestLog log) {
-        final ImmutableList.Builder<Tag> tagListBuilder = ImmutableList.builderWithExpectedSize(3);
+        final ImmutableList.Builder<Tag> tagListBuilder = ImmutableList.builderWithExpectedSize(2);
```
Thanks. 😄

ikhoon replied:
It is trivial, however, if we want to optimize this:

```java
int expectedSize = serviceTagName != null ? 3 : 2;
ImmutableList.Builder<Tag> tagListBuilder = ImmutableList.builderWithExpectedSize(expectedSize);
```

minwoox replied:
It's only two tags in this case, so I think it's `2`. 😄

ikhoon replied:
I misread the code.

arhont375 commented:
Rebased to fix failed checks, also I moved `RetrofitMeterIdPefixFunctionTest` from `.meter` package as no other classes there including just `RetrofitMeterIdPrefixFunction` and now it's late to change as it would be breaking change

arhont375 commented on retrofit2/src/main/java/com/linecorp/armeria/client/retrofit2/RetrofitMeterIdPrefixFunctionBuilder.java:
```diff
@@ -33,28 +33,77 @@
+     * @param serviceClass class that defines retrofit client service.
+     */
+    public RetrofitMeterIdPrefixFunctionBuilder serviceClass(Class<?> serviceClass) {
+        this.serviceClass = serviceClass;
```
Forgot to add here `requireNonNull` 😢

arhont375 commented on retrofit2/src/main/java/com/linecorp/armeria/client/retrofit2/RetrofitClassAwareMeterIdPrefixFunction.java:
```diff
@@ -0,0 +1,174 @@
+            final Method valueMethod;
+            try {
+                valueMethod = annotation.getClass().getMethod("value");
+                httpPath = (String) valueMethod.invoke(annotation);
```
I was thinking about switch, but it would look ugly also increase repetition of code that will may lead to errors 🌵

arhont375 commented:
Rebased, fixed after merge.

trustin commented:
Thanks a lot for your patience, @arhont375 !
